File: Army/mystic/spellcaster.cpp
#include "spellcaster.h"

using namespace std;

Spellcaster::Spellcaster(const string& name, int healthPoint, int damage, int manaPoint) : Unit(name, healthPoint, damage) {
    this->MPLimits = manaPoint;
    this->currentMP = manaPoint;
    learnSpell(new Fireball);
    learnSpell(new Heal);
}

Spellcaster::~Spellcaster() {
    cout << spellbook.size() << endl;
    spellbook.clear();
    cout << spellbook.size() << endl;
}

void Spellcaster::castSpell(Unit* victim) {
    ensureIsAlive();
    ensureIsNotSelfAttack(victim);
    ensureManaIsNotOver();
    
    getAbility()->castSpell(victim, currentSpell); 
    currentMP -= getCurrentSpell()->getManaConsumption();
}

void Spellcaster::learnSpell(Spell* newSpell) {
    spellbook[newSpell->getSpellsName()] = newSpell;
}

Spell* Spellcaster::getCurrentSpell() const {
    return currentSpell;
}

int Spellcaster::getMPLimit() const {
    return MPLimits;
}

int Spellcaster::getCurrentMP() const {
    return currentMP;
}

void Spellcaster::setMPLimit(int newMPLimit) {
    MPLimits = newMPLimit;
}

void Spellcaster::setCurrentMP(int newCurrentMP) {
    currentMP = newCurrentMP;
}

void Spellcaster::setCurrentSpell(const string& newCurrentSpell) {
    currentSpell = spellbook[newCurrentSpell];
}

void Spellcaster::ensureManaIsNotOver() {
    if ( getCurrentMP() < getCurrentSpell()->getManaConsumption() ) {
        throw ManaIsOverException();
    }
}

ostream& operator<<(ostream& out, const Spellcaster& spellcaster) {
    out << spellcaster.getName() << "\nHP: " << spellcaster.getHPLimit() << "|" << spellcaster.getCurrentHP() << endl;
    if ( spellcaster.getHPLimit() == 0 ) {
        return out;
    }
    out << "MP: " << spellcaster.getMPLimit() << "|" << spellcaster.getCurrentMP() << endl;
    out << "DMG: " << spellcaster.getDamage() << endl;

    return out;
}

File: Army-Java/Spellcasters/Spellcaster.java
package com.gymfox.Army.Spellcasters;

import com.gymfox.Army.MagicSkills.MagicSkills;
import com.gymfox.Army.Spells.Fireball;
import com.gymfox.Army.Spells.Heal;
import com.gymfox.Army.Spells.Spell;
import com.gymfox.Army.Units.Unit;

import java.util.*;

public abstract class Spellcaster extends Unit {
    private int manaPointLimits;
    private int currentMP;
    private Spell currentSpell;
    private MagicSkills magicPower;
    private Map<String, Spell> spellbook = new HashMap<>();
    private static final String DEFAULT_SPELL = "Fireball";
    private final static List<Spell> DEFAULT_SPELL_BOOK = Collections.unmodifiableList(Arrays.asList(new Fireball(),
            new Heal()));

    public static class ManaIsOverException extends Exception{}

    public Spellcaster(String name, int healthPointLimit, int damage, int manaPointLimits, MagicSkills magicPower,
                       List<Spell> DEFAULT_SPELL_BOOK) {
        this(name, healthPointLimit, damage, manaPointLimits, magicPower, DEFAULT_SPELL_BOOK, DEFAULT_SPELL);
    }

    public Spellcaster(String name, int healthPointLimit, int damage, int manaPointLimits, MagicSkills magicPower,
                       String spell) {
        this(name, healthPointLimit, damage, manaPointLimits, magicPower, DEFAULT_SPELL_BOOK, spell);
    }

    public Spellcaster(String name, int healthPointLimit, int damage, int manaPointLimits, MagicSkills magicPower,
                       List<Spell> spells, String currentSpell) {
        super(name, healthPointLimit, damage);
        this.manaPointLimits = manaPointLimits;
        this.currentMP = manaPointLimits;
        learnAllSpells(spells);
        this.magicPower = magicPower;
        this.currentSpell = spellbook.get(currentSpell);
    }

    private void learnAllSpells(List<Spell> spells) {
        for (Spell spell : spells) {
            spellbook.put(spell.getSpellsName(), spell);
        }
    }

    public void castSpell(Unit victim) throws IsSelfAttackException, UnitIsDeadException, ManaIsOverException {
        ensureIsNotSelfAttack(victim);
        ensureIsAlive();
        ensureManaIsNotOver();

        getCurrentSpell().applySpell(victim, magicPower);

        setCurrentMP(getCurrentMP() - getCurrentSpell().getManaConsumption());
    }

    protected void ensureManaIsNotOver() throws ManaIsOverException {
        if ( getCurrentMP() < getCurrentSpell().getManaConsumption() ) {
            throw new ManaIsOverException();
        }
    }

    protected void learnSpell(Spell newSpell) {
        spellbook.put(newSpell.getSpellsName(), newSpell);
    }

    public int getManaPointLimits() {
        return manaPointLimits;
    }

    public int getCurrentMP() {
        return currentMP;
    }

    public Spell getCurrentSpell() {
        return currentSpell;
    }

    public void setCurrentMP(int newCurrentMP) {
        this.currentMP = newCurrentMP;
    }

    public void setCurrentSpell(String newCurrentSpell) {
        currentSpell = spellbook.get(newCurrentSpell);
    }
}
